xrdfs rm - not exiting with a !=0 exit code in case of errors

Hi,

Since I can't reopen https://github.com/xrootd/xrootd/issues/1885 I'll copy-paste the request here.

This bug seems to have reappeared at least from 5.6.0 and beyond. Even for single files, regardless of the success or failure of the operation, the exit code is always 0:

```
$ v5.6.0-6/bin/xrdfs eosalice.cern.ch rm '/33/33333/33333'
rm /33/33333/33333 : [ERROR] Error response: permission denied
$ echo $?
0

$ v5.6.0-6/bin/xrdfs eosalice.cern.ch rm '/33/33333/33333' '/44/44444/444444'
rm /33/33333/33333 : [ERROR] Error response: permission denied
rm /44/44444/444444 : [ERROR] Error response: permission denied
$ echo $?
0
```

and the same in 5.6.2:

```
$ xrdcp -version
v5.6.2
$ xrdfs eosalice.cern.ch rm '/33/33333/33333'
rm /33/33333/33333 : [ERROR] Error response: permission denied
$ echo $?
0
```

Could you please re-apply the fix?

Cheers,

.costin

--- a/src/XrdCl/XrdClFS.cc
+++ b/src/XrdCl/XrdClFS.cc
@@ -701,9 +701,12 @@ XRootDStatus DoRm( FileSystem                      *fs,
   }
 
   //----------------------------------------------------------------------------
-  // Run the query
+  // Run the query:
+  // Parallel() will take the vector of Pipeline by reference and empty the
+  // vector, so rms.size() will change after the call.
   //----------------------------------------------------------------------------
-  XRootDStatus st = WaitFor( Parallel( rms ).AtLeast( rms.size() ) );
+  const size_t rs = rms.size();
+  XRootDStatus st = WaitFor( Parallel( rms ).AtLeast( rs ) );
   if( !st.IsOK() )
     return st;
 

--- a/tests/XRootD/smoke.sh
+++ b/tests/XRootD/smoke.sh
@@ -99,6 +99,45 @@ done
 
 wait
 
+#
+# check return code for xrdfs rm
+# create another 6 files, which should be deleted during the test
+#
+for i in $(seq -w 1 6) ; do
+       ${OPENSSL} rand -out "${TMPDIR}/${i}.exists.ref" $((1024 * $RANDOM))
+       ${XRDCP} ${TMPDIR}/${i}.exists.ref ${HOST}/${TMPDIR}/${i}.exists.ref
+done
+
+# remove 3 existing, should succeed not error
+${XRDFS} ${HOST} rm ${TMPDIR}/1.exists.ref ${TMPDIR}/2.exists.ref ${TMPDIR}/3.exists.ref
+
+set +e
+
+# remove 3 not existing, should error
+#
+${XRDFS} ${HOST} rm ${TMPDIR}/not_exists_1.ref ${TMPDIR}/not_exists_2.ref ${TMPDIR}/not_exists_3.ref
+rm_rc=$?
+if [ $rm_rc -eq 0 ]; then
+  exit 1
+fi
+#
+# remove 2 existing, 1 not existing should error
+#
+${XRDFS} ${HOST} rm ${TMPDIR}/4.exists.ref ${TMPDIR}/5.exists.ref ${TMPDIR}/not_exists_4.ref
+rm_rc=$?
+if [ $rm_rc -eq 0 ]; then
+  exit 1
+fi
+#
+# remove 1 existing, 2 not existing should error
+#
+${XRDFS} ${HOST} rm ${TMPDIR}/6.exists.ref ${TMPDIR}/not_exists_5.ref ${TMPDIR}/not_exists_6.ref
+rm_rc=$?
+if [ $rm_rc -eq 0 ]; then
+  exit 1
+fi
+set -e
+
 ${XRDFS} ${HOST} rmdir ${TMPDIR}
 
 echo "ALL TESTS PASSED"

